### PR TITLE
extend documentation on server-side apply

### PR DESCRIPTION
**Pull request by kwiesmueller**
This PR adds more documentation for server-side apply and serves as the placeholder for potential other additions that are targeted for 1.15.

Current additions:
- Documentation on how to clear the managedFields object manually.

/milestone 1.15
/sig api-machinery
/wg apply
/cc @apelisse @jennybuckley @simplytunde

### CODE REVIEW

**Comment by tengqm** [comment]
Please indicate if this is a WIP.

**Comment by kwiesmueller** [comment]
sorry @tengqm 
yes it's wip for now.

File: content/en/docs/reference/using-api/api-concepts.md [hunk 1]
```diff
@@ -430,3 +430,17 @@ Another difference is that an applier using Client Side Apply is unable to chang
 ### Custom Resources
 
 Server Side Apply currently treats all custom resources as unstructured data. All keys are treated the same as struct fields, and all lists are considered atomic. In the future, it will use the validation field in Custom Resource Definitions to allow Custom Resource authors to define how to how to merge their own objects.
+
+### Clearing ManagedFields
+
+It is possible to strip all managedFields from an object by overwriting them using `MergePatch`.
```
Comment (apelisse) [inline content/en/docs/reference/using-api/api-concepts.md]: does it only work with `mergepatch`?

**Comment by apelisse** [comment]
I think this is still relevant for 1.15. @kwiesmueller would you mind following-up on that?

**Comment by kwiesmueller** [comment]
Sure, I'll take care of it.

File: content/en/docs/reference/using-api/api-concepts.md [hunk 2]
```diff
@@ -430,3 +430,24 @@ Another difference is that an applier using Client Side Apply is unable to chang
 ### Custom Resources
 
 Server Side Apply currently treats all custom resources as unstructured data. All keys are treated the same as struct fields, and all lists are considered atomic. In the future, it will use the validation field in Custom Resource Definitions to allow Custom Resource authors to define how to how to merge their own objects.
+
+### Clearing ManagedFields
+
+It is possible to strip all managedFields from an object by overwriting them using `MergePatch` or `JSONPatch`.
```
Comment (apelisse) [inline content/en/docs/reference/using-api/api-concepts.md]: What about smp, what about updates?

**Comment by kbhawkey** [comment]
Page preview:
https://5cf67bce00c2540008f36c97--kubernetes-io-vnext-staging.netlify.com/docs/reference/using-api/api-concepts/#server-side-apply

**Comment by jimangel** [comment]
@apelisse @apelisse @jennybuckley @simplytunde @kwiesmueller: Can I get a technical /lgtm? Otherwise this looks good...

/approved

**Comment by kwiesmueller** [comment]
@jimangel I'll see to respond to @apelisse last question. I think the other options should work the same, but didn't test it (and don't know how many examples we want for this).

**Comment by makoscafee** [comment]
Hi @kwiesmueller, sorry can you point me to the enhancement this PR is providing docs for? Since i cannot find it in enhancement sheet, so should not be forgotten. Thanks

**Comment by kwiesmueller** [comment]
@makoscafee the feature targeted by this PR is introduced in this KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/0006-apply.md

**Comment by makoscafee** [comment]
@kwiesmueller  Perfect, its just that this KEP is targeted for 1.16 I think https://github.com/kubernetes/enhancements/issues/555#issuecomment-496602774. Correct me if I am wrong. It did not make it to 1.15 release cut.

**Comment by kwiesmueller** [comment]
@makoscafee the feature we talk about is already in alpha since 1.14. This PR got added to address a question as part of our umbrella issue https://github.com/kubernetes/kubernetes/issues/73723 to show people how to reset a field that got introduced in 1.14 (alpha).

So adding this is only related to the feature inside the KEP but directly targets the current release (1.15) on purpose.
The test proving the documentation added here works as intended got added here: https://github.com/kubernetes/kubernetes/pull/77821

**Comment by makoscafee** [comment]
Thanks for clarification @kwiesmueller. Do you think will it be possible for it to be merged by today? we are behind schedule on getting all PR for this release merged.

Let me know if there is anything needed.

/hold clear

**Comment by kwiesmueller** [comment]
@apelisse would it be okay for you to merge this?
I think all other options (except apply) should work but couldn't test them yet. If we wan't to document all the options, I'd prefer having tests for them as well and then add them to the docs which won't happen in this release due to code freeze (i think).

**Comment by kwiesmueller** [comment]
@apelisse I updated the PR and added this one for https://github.com/kubernetes/kubernetes/pull/78912 verifying the docs work in all cases.

**Comment by apelisse** [comment]
Thanks a lot Kevin, looking at it now (I ended-up in various meeting this morning so I couldn't exactly do it when I said I would)

**Comment by apelisse** [comment]
@jimangel Did you mean `/approve`? :-)

**Comment by apelisse** [comment]
One thing that is not super clear is ... what does it mean to strip all the managed fields? what about the operation that you're running as you clear?

Like.. what if I do:
```json
PATCH /api/v1/namespaces/default/configmaps/example-cm
Content-Type: application/merge-patch+json
Accept: application/json
Data: {"metadata":{"managedFields": [{}]}, "other": "changes"}
```

What is `managedField` going to look like after this patch? Will it be empty or will it be me owning `other`?

**Comment by kwiesmueller** [comment]
I just tested this, seems like when you overwrite the managedFields in a request, this is the final operation. So in your example, the managedFields are empty.
We really got to document this, so I'll add it. But maybe we also should discuss if this is how we want it to behave?

**Comment by apelisse** [comment]
I agree that seems like a bug to me.

**Comment by makoscafee** [comment]
Hi, @kwiesmueller and @apelisse so since this PR has some work to be done. and the deadline to have all PR passed since June 10th.I am going to strip it from 1.15 milestone and we can get it in as a fast follow -- retarget for master after the release, since we are publishing continuously from master.

**Comment by makoscafee** [comment]
Closing this PR as it is referencing `dev-1.15`and the cycle has ended.

**Pull request closed without merging**